slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -329,6 +332,109 @@ void SLAMServiceImpl::SetUpMapBuilder() {
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+
+    // total number of bytes, 32 bits per pixel
+    size_t size_data = width * height * 4;
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+
+    std::stringbuf data_buffer;
+
+    int num_points = 0;
+
+    // Reduce resolution based off number of pixels. Output is pixels to skip
+    int scaleFactor = 4;
+    int scale = size_data / 100000.f / 32.f / scaleFactor;
+
+    // Loop to filter unwanted data and reduce resolution
+    for (int i = 0; i < size_data; i = i + scale * 4) {
+        int rgb = 0;
+        rgb = rgb | ((int)data_vect[i + 0] << 16);
+        rgb = rgb | ((int)data_vect[i + 1] << 8);
+        rgb = rgb | ((int)data_vect[i + 2] << 0);
+
+        // skip pixels that are not in our map(black/past walls)
+        //  this value represents [102,102,102]
+        if (rgb == 6710886) continue;
+
+        // Determine probability based off color pixel
+        int prob = ViamColorToProbability((int)data_vect[i + 2]);
+
+        if (prob == 0) continue;
+
+        num_points++;
+
+        int pixel_index = i / 4;
+        int pixel_x = pixel_index % width;
+        int pixel_y = pixel_index / width;
+
+        // kPixelSize is .01, unsure of units but assumed converts to meters
+        // based on PaintSubmapSlices() and DrawPoseOnSurface()
+        float x_pos = (pixel_x - painted_slices.origin.x()) * kPixelSize;
+        // Y is inverted to match output from getPosition()
+        float y_pos = -(pixel_y - painted_slices.origin.y()) * kPixelSize;
+        // 2D SLAM so Z is set to 0
+        float z_pos = 0;
+
+        // rotating coordinates to match slam service expectation(XZ plane)
+        data_buffer.sputn((const char *)&x_pos, 4);
+        data_buffer.sputn((const char *)&z_pos, 4);
+        data_buffer.sputn((const char *)&y_pos, 4);
+        data_buffer.sputn((const char *)&prob, 4);
+    }
+
+    std::stringbuf pointcloud_buffer;
+    std::ostream oss(&pointcloud_buffer);
+
+    // Write our PCD file, which is written as a binary.
+    oss << "VERSION .7\n"
+        << "FIELDS x y z rgb\n"
+        << "SIZE 4 4 4 4\n"
+        << "TYPE F F F I\n"
+        << "COUNT 1 1 1 1\n"
+        << "WIDTH " << num_points << "\n"
+        << "HEIGHT " << 1 << "\n"
+        << "VIEWPOINT 0 0 0 1 0 0 0\n"
+        << "POINTS " << num_points << "\n"
+        << "DATA binary\n";
+
+    // writes data buffer to the pointcloud string
+    oss << data_buffer.str();
+    pointcloud = pointcloud_buffer.str();
+    return true;
+}
+
+int SLAMServiceImpl::ViamColorToProbability(int color) {
+    int maxVal = 255;
+    int minVal = 102;
+    int maxProb = 100;
+    int minProb = 0;
+    int prob =
+        (color - maxVal) * (maxProb - minProb) / (minVal - maxVal) + minProb;
```

Comment:
[q] Is it worth also calculating the minVal and maxVal from the image? It would require going over the image twice. Once to find max/min and once to perform this calculation for each one so may not be worth it.

Note: Just as an FYI, the parallel projection code, a negative value or value > 100 will be ignored. Still a good check to do.

Reply:
maxVal and minVal are currently hardcoded based off visual inspection. I am confident the values will not change between runs, but I am unsure of how they will change if cairo/painting parameters are adjusted